lib/evmone/eof.cpp

```diff
@@ -349,40 +349,41 @@ bool validate_rjump_destinations(bytes_view code) noexcept
 std::variant<EOFValidationError, int32_t> validate_max_stack_height(
     bytes_view code, size_t func_index, const std::vector<EOFCodeType>& code_types)
 {
-    assert(!code.empty());
-
     // Special values used for detecting errors.
     static constexpr int32_t LOC_UNVISITED = -1;  // Unvisited byte.
-    static constexpr int32_t LOC_IMMEDIATE = -2;  // Immediate byte.
 
     // Stack height in the header is limited to uint16_t,
     // but keeping larger size for ease of calculation.
-    std::vector<int32_t> stack_heights(code.size(), LOC_UNVISITED);
-    stack_heights[0] = code_types[func_index].inputs;
+    struct StackHeightRange
+    {
+        int32_t min = LOC_UNVISITED;
+        int32_t max = LOC_UNVISITED;
+    };
 
-    std::stack<size_t> worklist;
-    worklist.push(0);
+    assert(!code.empty());
 
-    while (!worklist.empty())
-    {
-        const auto i = worklist.top();
-        worklist.pop();
+    std::vector<StackHeightRange> stack_heights(code.size(), {LOC_UNVISITED, LOC_UNVISITED});
+    stack_heights[0] = {code_types[func_index].inputs, code_types[func_index].inputs};
 
+    for (size_t i = 0; i < code.size();)
+    {
         const auto opcode = static_cast<Opcode>(code[i]);
 
         int stack_height_required = instr::traits[opcode].stack_height_required;
         auto stack_height_change = instr::traits[opcode].stack_height_change;
 
-        auto stack_height = stack_heights[i];
-        assert(stack_height != LOC_UNVISITED);
+        const auto stack_height = stack_heights[i];
+        if (stack_height.min == LOC_UNVISITED)
+            return EOFValidationError::unreachable_instructions;
```

Comment:
This is tricky. Please explain in a comment.

Reply:
Related to https://github.com/ethereum/evmone/pull/676/files/f424a769d99d8fd96b0f5d908a8194ace20d3b39#r1496139888, maybe error name is not ideal, but not quite sure how to improve.

I'll add some comment.